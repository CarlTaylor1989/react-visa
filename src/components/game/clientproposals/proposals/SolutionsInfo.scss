@import '../../../../styles/generic';
@import '../../../../styles/popup';

$faded-border-colour: rgba($colour-white, .5);
$feedback-panel-width: 450px;
$popup-close-n: '../../../../styles/icons/popup_close_n.svg';
$popup-close-h: '../../../../styles/icons/popup_close_h.svg';
$popup-close-height: 50px;
$popup-close-width: 50px;

.solutionsInfo {
  @extend %gradient-popup;
  @include border-radius(27px);
  @include box-shadow(21px, 21px, 11px, 0, rgba(0, 0, 0, .3));
  @include opacity(0);
  border: 2px solid $faded-border-colour;
  box-sizing: border-box;
  height: 665px;
  padding: 100px 30px 30px;
  position: absolute;
  right: 0;
  top: 10px;
  transition: opacity .5s ease-in-out;
  width: $feedback-panel-width;
  z-index: -1;

  &.available {
    @include opacity(1);
    z-index: 10;
  }

  .exitPopup {
    @extend %un-button;
    background-image: url($popup-close-n);
    background-size: $popup-close-width $popup-close-height;
    height: $popup-close-height;
    position: absolute;
    right: 25px;
    top: 25px;
    transition: background .5s ease-in-out;
    width: $popup-close-width;
    z-index: 2;

    &:hover {
      background-image: url($popup-close-h);
      background-size: $popup-close-width $popup-close-height;
    }
  }

  .popupWrapper {
    padding-right: 20px;

    .solutionTitle,
    .solutionSubtitle,
    .solutionText {
      display: block;
      text-align: left;
    }

    .solutionTitle {
      color: $colour-gold;
      font-size: 36px;
    }

    .solutionSubtitle {
      color: $colour-gold;
      font-size: 28px;
      margin-top: 20px;
    }

    .solutionText {
      font-size: 18px;
      margin-top: 20px;
    }
  }

  .contentScroll {
    @extend %scrollbar;
    height: 525px;

    .resourceWrapper {
      padding-right: 30px;
    }
  }
}
