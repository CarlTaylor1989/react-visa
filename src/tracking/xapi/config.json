{
  "version": "1.0.0",
  "projectId": "mario19-037",
  "gameName": "visa-inside-track",
  "baseURL": "https://lms.visa.com/e-learning/",
  "extensionURL": "https://lms.visa.com/xapi/cornerstone/extensions/",
  "title": "Mario technical prototype",
  "homePage": "https://lms.visa.com",
  "gomoCourseURL": "http://www.gomolearning.com/xapi/activities/course/", 
  "debugMode": false
}
