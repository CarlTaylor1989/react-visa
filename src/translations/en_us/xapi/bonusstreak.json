{
  "title": "Bonus Streak",
  "description": "A bonus streak awarded for successfully completing for the first time four challenges in a row."
}
