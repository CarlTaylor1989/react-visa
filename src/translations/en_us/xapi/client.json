{
  "questionTitle": "%{name} client proposal step %{index}",
  "questionDescription": "%{name} client proposal with topic %{problem}, %{question} selection:",
  "question1": "statement",
  "question2": "product",

  "completedDescription": "A client proposal with topic %{problem}"
}
