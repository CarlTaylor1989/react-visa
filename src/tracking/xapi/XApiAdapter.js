import TinCan from 'tincanjs';
import _ from 'lodash';
import XApiRegistry from './XApiRegistry';
import config from './config.json';
import { cleanupText } from '../../lib/htmlutils';

const langauge = process.env.REACT_APP_LANGUAGE || 'en_us';
const xApiLang = process.env.XAPI_LANGUAGE || 'en-US';
const content = require(`../../translations/${langauge}/xapi/index`).default; // eslint-disable-line import/no-dynamic-require

const uuidv4 = require('uuid/v4');

let lrsInstance = null;
let startTime = null;
let agentAccount;
const interactionTypes = [
  'truefalse',
  'choice',
  'fill-in',
  'long-fill-in',
  'matching',
  'performance',
  'sequencing',
  'likert',
  'numeric',
  'other'
];

export const lib = {};

lib.log = (msg) => {
  if (config.debugMode) {
    console.log(msg); // eslint-disable-line no-console
  }
};

/**
 * Returns the endpoint details based on the platform
 * @return {object}
 */
lib.getEndpointDetails = () => {
  let endpointDetails;
  if (process.env.PLATFORM === 'local') {
    endpointDetails = {
      endpoint: process.env.DEV_ENDPOINT_URL,
      username: process.env.DEV_ENDPOINT_KEY,
      password: process.env.DEV_ENDPOINT_SECRET
    };
  } else if (process.env.PLATFORM === 'test') {
    endpointDetails = {
      endpoint: process.env.LEO_TEST_ENDPOINT_URL,
      username: process.env.LEO_TEST_ENDPOINT_KEY,
      password: process.env.LEO_TEST_ENDPOINT_SECRET
    };
  } else if (process.env.MODE === 'test') { // production + test
    endpointDetails = {
      endpoint: process.env.VISA_TEST_ENDPOINT_URL,
      username: process.env.VISA_TEST_ENDPOINT_KEY,
      password: process.env.VISA_TEST_ENDPOINT_SECRET
    };
  } else { // production + live
    endpointDetails = {
      endpoint: process.env.VISA_LIVE_ENDPOINT_URL,
      username: process.env.VISA_LIVE_ENDPOINT_KEY,
      password: process.env.VISA_LIVE_ENDPOINT_SECRET
    };
  }
  return endpointDetails;
};

/**
 * Create agent account definition
 * @param {object} learner The learner email and name
 * @return {object}
 */
lib.createAgentAccount = (learner) => {
  const agent = {};
  const uuid = uuidv4();
  if (learner && learner.id) {
    if (learner.id.includes('@')) {
      agent.mbox = learner.id;
    } else {
      agent.account = {
        homePage: config.homePage,
        name: learner.id
      };
    }
  } else {
    agent.mbox = `${uuid}.xapiuser@leolearning.com`;
  }
  if (learner && learner.name) {
    agent.name = learner.name;
  } else {
    agent.name = `User: ${uuid}`;
  }
  return agent;
};

/**
 * Create course activity object
 * @return {object}
 */
lib.getMainActivity = () => ({
  type: 'course',
  id: config.gameName,
  title: content.course.title,
  description: content.course.description
});

/**
 * Get browser info from the navigator
 * @return {object}
 */
lib.getBrowserInfo = () => ({
  code_name: navigator.appCodeName,
  name: navigator.appName,
  version: navigator.appVersion,
  platform: navigator.platform,
  'user-agent-header': navigator.userAgent,
  'cookies-enabled': navigator.cookieEnabled
});

/**
 * Parse URSL and extract the parameters
 * @param {string} url URL to parse
 */
lib.parseURL = (url) => {
  // Create result object
  const result = {
    url: '',
    params: {},
    original: url
  };
  // Parse the URL
  if (url.includes('?')) {
    const parts = url.split('?');
    result.url = parts[0];
    const params = parts[1].split('&');
    for (let x = 0; x < params.length; x += 1) {
      const param = params[x].split('=');
      result.params[param[0]] = param[1];
    }
  }
  // Return the results object
  return result;
};

/**
 * Replace delimiters with correct xapi ones.
 * @param {string} value
 */
lib.formatDelimiters = (value) => {
  let formatted = value;
  formatted = formatted.replace(/,/g, '[,]');
  formatted = formatted.replace(/\./g, '[.]');
  return formatted;
};

/**
 * Get launched URL and protocol
 * @param {object} location The window location object
 * @return {object}
 */
lib.getLaunchedURL = (location) => {
  let url = location.toString();
  const urlParts = url.split(':');
  let protocol = 'http';
  if (url && urlParts.length) {
    protocol = urlParts.shift().toLowerCase();
    protocol = (protocol === 'http' || protocol === 'https' ? protocol : 'http');
    url = `${protocol}:${urlParts.join(':')}`;
  }
  return {
    url,
    protocol
  };
};

/**
 * Create verb object
 * @param {string} verb Verb name
 * @return {object}
 */
lib.createVerb = (verb) => {
  const registeredVerb = XApiRegistry.verbs[verb];
  if (registeredVerb === undefined) {
    lib.log(`Verb ${verb} not in the registry`);
    return '';
  }
  return new TinCan.Verb({
    id: registeredVerb.id,
    display: {
      [xApiLang]: registeredVerb.name
    }
  });
};

/**
 * Create activity object
 * @param {string} activity Activity name
 * @param {object} extra Override activity definition
 * @return {object}
 */
lib.createActivity = (activity, extra) => {
  let newActivity = activity;
  const newExtra = extra;
  if (typeof activity === 'string') {
    newActivity = { type: activity };
  } else {
    newActivity = activity || {};
  }

  let activityId = '';
  let activityDefinition = {};

  // Set the registered activity details...
  const registeredActivity = XApiRegistry.activities[newActivity.type];
  if (registeredActivity !== undefined) {
    activityId = registeredActivity.id;
    activityDefinition.name = {
      [xApiLang]: registeredActivity.name
    };
    activityDefinition.type = registeredActivity.id;
  }

  if (!newActivity.id && newActivity.type === 'course') {
    newActivity = Object.assign(newActivity, lib.getMainActivity());
  }

  // Override with any user specified.
  if (newActivity && newActivity.id && newActivity.type) {
    if (newActivity.type === 'course' && !newActivity.gomoCourse) {
      activityId = config.baseURL + newActivity.id;
    } else if (newActivity.type === 'course' && newActivity.gomoCourse) {
      activityId = config.gomoCourseURL + newActivity.id;
    } else {
      activityId = `${config.baseURL}${config.gameName}/${newActivity.id}`;
    }
  }

  // Override with any user specified.
  if (newActivity.title !== undefined && newActivity.title) {
    activityDefinition.name = {
      [xApiLang]: cleanupText(newActivity.title)
    };
  }

  if (newActivity.description !== undefined && newActivity.description) {
    activityDefinition.description = {
      [xApiLang]: cleanupText(newActivity.description)
    };
  }

  if (newActivity.extensions) {
    activityDefinition.extensions = newActivity.extensions;
  }

  if (newActivity.type && interactionTypes.includes(newActivity.type)) {
    // @see {@link https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Data.md#requirements-4}
    activityDefinition.type = XApiRegistry.activities['cmi.interactions'].id;
    // @see {@link https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Data.md#details-10}
    activityDefinition.interactionType = newActivity.type;
  }

  // Format any question options and reponses
  if (newActivity.correctResponses && _.isArray(newActivity.correctResponses)) {
    newActivity.correctResponses = newActivity.correctResponses.map(cr => lib.formatDelimiters(cr));
  }

  if (newExtra && newExtra.result) {
    if (newExtra.result.response && _.isString(newExtra.result.response)) {
      newExtra.result.response = lib.formatDelimiters(newExtra.result.response);
    }

    if (newExtra.result.extensions) {
      newExtra.result.extensions = {
        [config.extensionURL + newExtra.result.extensions.name]: newExtra.result.extensions.value
      };
    }
  }


  if (newActivity.choices) {
    activityDefinition.choices = newActivity.choices.map(choice => ({
      id: choice.id,
      description: {
        [xApiLang]: cleanupText(choice.description)
      }
    }));
  }

  if (newActivity.correctResponses) {
    activityDefinition.correctResponsesPattern = newActivity.correctResponses;
  }

  // Override definition if a statement has defined its own.
  if (newExtra && newExtra.object && newExtra.object.definition) {
    activityDefinition = newExtra.object.definition;
  }

  if (activityId && activityDefinition) {
    // Create the activity
    return new TinCan.Activity({
      id: activityId,
      definition: activityDefinition
    });
  }

  lib.log('Failed to create activity');
  return null;
};

/**
 * Create agent object
 * @return {object}
 */
lib.createAgent = (agent) => {
  let xapiAgent = null;
  if (_.isObject(agent) && !_.isEmpty(agent)) {
    xapiAgent = new TinCan.Agent(agent);
  }
  return xapiAgent;
};

lib.getLRSInstance = () => lrsInstance;

lib.setLRSInstance = (lrs) => {
  let result = false;
  if (lrs && !lrsInstance) {
    lrsInstance = lrs;
    result = true;
  }
  return result;
};

/**
 * Create result object
 * @param {object} resultObj
 */
lib.createResult = (resultObj) => {
  const result = resultObj || {};

  if (result.score && _.isObject(result.score)) {
    result.score.raw = Number.isNaN(+result.score.raw) ? 0 : +result.score.raw;
    // Create a TinCan score object
    result.score = new TinCan.Score(result.score);
  }

  // Create and return TinCan result object
  return new TinCan.Result(result);
};

/**
 * Send xapi statement
 * @param {string} verb Statement verb
 * @param {string} activity Statement activity name
 * @param {object} extra Activity definition override
 */
lib.sendStatement = (verb, activity, extra) => {
  const lrs = lib.getLRSInstance();
  if (!lrs) return;

  const tcVerb = lib.createVerb(verb);
  if (typeof tcVerb !== 'object') {
    return;
  }

  // Create statement definition
  const statementDefinition = {
    actor: lib.createAgent(agentAccount || lib.createAgentAccount()),
    verb: tcVerb,
    target: lib.createActivity(activity, extra),
    version: config.version,
    context: {}
  };

  // Add any context information
  if (extra && extra.context) {
    statementDefinition.context = extra.context;
  }

  // Add any result information
  if (extra && extra.result && !_.isEmpty(extra.result)) {
    statementDefinition.result = lib.createResult(extra.result);
  }

  if (!statementDefinition.context.contextActivities) {
    statementDefinition.context.contextActivities = {};
  }

  // Add grouping to all statements
  statementDefinition.context.contextActivities.grouping = [{
    id: config.baseURL + config.gameName,
    objectType: 'Activity'
  }];

  // Create statement and save it
  const statement = new TinCan.Statement(statementDefinition);

  lrs.saveStatement(
    statement,
    {
      callback: (err, xhr) => {
        if (err !== null) {
          if (xhr !== null) {
            lib.log(`Failed to save statement: ${xhr.responseText} (${xhr.status})`); // eslint-disable-line max-len, no-console
          } else {
            lib.log(`Failed to send statement ${err}`); // eslint-disable-line no-console
          }
        } else {
          lib.log('Statement saved'); // eslint-disable-line no-console
        }
      }
    }
  );
};

/**
 * Content launched or resumed
 * @param {boolean} resumed Course either resumed or launched for first time
 */
lib.sendLaunchedResumed = (resumed) => {
  const activity = lib.getMainActivity();
  const extra = {
    context: {
      extensions: {
        [XApiRegistry.extensions['browser-info'].id]: lib.getBrowserInfo()
      }
    }
  };

  lib.sendStatement(resumed ? 'resumed' : 'launched', activity, extra);
};

/**
 * Get state from the LRS
 * @param {string} stateId
 * @param {string} activity
 * @return {object}
 */
lib.getState = (stateId, activity) => {
  const lrs = lib.getLRSInstance();
  if (!lrs) return {};

  let result = lrs.retrieveState(stateId, {
    agent: lib.createAgent(agentAccount),
    activity: lib.createActivity(activity),
    contentType: 'application/json'
  });

  if (result && !result.err && result.state && result.state.contents) {
    result = result.state.contents;
  } else {
    result = {};
  }
  return result;
};

/**
 * Set state to the LRS
 * @param {string} stateId
 * @param {string} stateData
 * @param {string} activity
 * @param {function} callback
 */
lib.saveState = (stateId, stateData, activity, callback) => {
  const lrs = lib.getLRSInstance();
  if (!lrs) return;

  lrs.saveState(stateId, stateData, {
    agent: lib.createAgent(agentAccount),
    activity: lib.createActivity(activity),
    contentType: 'application/json',
    callback
  });
};

/**
 * Create an TinCan compatible duration from two date/time stamps.
 * @param {data} start
 * @param {data} end
 * @return {string}
 */
lib.createDuration = (start, end) => {
  let duration = '';
  if (start instanceof Date && end instanceof Date) {
    const msecs = end.getTime() - start.getTime();
    duration = TinCan.Utils.convertMillisecondsToISO8601Duration(msecs);
  }
  return duration;
};

/**
 * Parse result
 * The results object will conform to the one defined in TinCan v1.0.0 spec.
 * @param {object} result The result object to be formatted.
 * @return {object}
 */
lib.parseResult = (resultToParse) => {
  const result = resultToParse || {};
  // Score object
  if (result.score && _.isObject(result.score)) {
    result.score.raw = Number.isNaN(+result.score.raw) ? 0 : +result.score.raw;
  }
  // Success status
  if (result.success !== undefined) {
    result.success = result.success || false;
  }

  // Completion status
  if (result.completion !== undefined) {
    result.completion = result.completion || false;
  }

  return result;
};

/**
 * Create referrer extension
 * @param {string} referrer The referrer
 * @return {object}
 */
lib.createReferrerExtension = referrer => ({
  [XApiRegistry.extensions.referrer.id]: {
    id: `${config.baseURL}${config.gameName}/${referrer}`,
    objectType: 'Activity'
  }
});

/**
 * Parse the interaction data
 * @param {Object} data
 * @return {Object}
 */
lib.parseInteractionData = (data) => {
  const interactionData = _.merge(data || {});
  interactionData.type = interactionData.type || '';
  interactionData.activity = interactionData.activity || {};
  interactionData.activity.id = interactionData.activity.id || '';
  interactionData.activity.title = interactionData.activity.title || ''; // eslint-disable-line max-len
  interactionData.activity.type = interactionData.activity.type || '';
  interactionData.result = lib.parseResult(interactionData.result);

  const extra = {
    result: _.clone(interactionData.result)
  };

  // if a statement has explicitly defined object data then use it
  if (_.isObject(interactionData.object)) {
    extra.object = interactionData.object;
  }

  if (_.isObject(interactionData.response)) {
    // populate result.response from response.detail
    if (_.isString(interactionData.response.detail)) {
      extra.result.response = interactionData.response.detail;
    }

    // populate result.duration from response.timeAvailable and response.timeResponse
    if (_.isDate(interactionData.response.timeAvailable)
      && _.isDate(interactionData.response.timeResponse)) {
      extra.result.duration = lib.createDuration(
        interactionData.response.timeAvailable,
        interactionData.response.timeResponse
      );
    }
  }

  if (interactionData.referrer) {
    extra.context = {
      extensions: lib.createReferrerExtension(interactionData.referrer)
    };
  }

  if (_.has(interactionData, 'parent') || interactionData.other) {
    extra.context = extra.context || {};
    extra.context.contextActivities = {};
  }

  if (_.has(interactionData, 'parent')) {
    const parent = interactionData.parent ? `/${interactionData.parent}` : '';
    extra.context.contextActivities.parent = [{
      id: `${config.baseURL}${config.gameName}${parent}`,
      objectType: 'Activity'
    }];
  }

  if (interactionData.other && interactionData.other.type === 'gomo') {
    extra.context.contextActivities.other = [{
      id: config.gomoCourseURL + interactionData.other.projectId,
      objectType: 'Activity'
    }];
  }

  return {
    interactionData,
    extra
  };
};

/**
 * Send interaction data
 * @param {object} data
 */
export const sendInteractionData = (data) => {
  const parsedData = lib.parseInteractionData(data);
  lib.sendStatement(
    parsedData.interactionData.type,
    parsedData.interactionData.activity,
    parsedData.extra
  );
};
lib.sendInteractionData = sendInteractionData;

/**
 * Create LRS store object
 * @param {object} data LRS config
 */
lib.createLRSStore = () => {
  const endpointConfig = lib.getEndpointDetails();
  const currentStore = {
    ...endpointConfig
  };

  const launchedUrl = lib.getLaunchedURL(window.location.toString());

  // Add the protocol if necessary
  if (currentStore.endpoint && /^\/\//.test(currentStore.endpoint)) {
    currentStore.endpoint = `${launchedUrl.protocol}:${currentStore.endpoint}`;
  }

  const parsedURL = lib.parseURL(currentStore.endpoint);
  if (parsedURL.url) {
    currentStore.endpoint = parsedURL.url;
    currentStore.extended = parsedURL.params;
  }

  return currentStore;
};

/**
 * Initialise LRS
 * @param {object} learner The learner name and email
 * @param {boolean} resumed Content resumed or launched for first time
 */
export const initialiseLrs = (learner, resumed) => {
  let lrs = lib.getLRSInstance();
  if (!lrs) {
    const currentStore = lib.createLRSStore();

    try {
      lrs = new TinCan.LRS({
        ...currentStore,
        allowFail: false
      });
      const result = lib.setLRSInstance(lrs);
      if (result) {
        lib.log('Connection established'); // eslint-disable-line no-console
      }

      if (config.debugMode) {
        TinCan.enableDebug();
      }

      startTime = new Date();
      agentAccount = lib.createAgentAccount(learner);
      lib.sendLaunchedResumed(resumed);
    } catch (e) {
      lib.log(`Failed to setup LRS object: ${JSON.stringify(e)}`); // eslint-disable-line no-console
    }
  }
};

/**
 * Terminate communication with the LRS
 */
export const exitLrs = () => {
  const endTime = new Date();
  const duration = lib.createDuration(startTime, endTime);
  const courseActivity = lib.getMainActivity();

  // Add duration to the result block
  const extra = {
    result: { duration }
  };

  lib.sendStatement('exited', courseActivity, extra);
  lrsInstance = null;
};
