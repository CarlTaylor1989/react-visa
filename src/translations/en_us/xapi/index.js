import course from './course.json';
import diagnostic from './diagnostic.json';
import introseq from './introseq.json';
import bonusstreak from './bonusstreak.json';
import ranks from './ranks.json';
import settings from './settings.json';
import map from './map.json';
import products from './products.json';
import dashboard from './dashboard.json';
import gomo from './gomo';
import clientproposals from './clientproposals.json';
import client from './client.json';
import tutorials from './tutorials.json';
import feedback from './feedback.json';

export default {
  course,
  diagnostic,
  introseq,
  bonusstreak,
  ranks,
  settings,
  map,
  products,
  dashboard,
  gomo,
  clientproposals,
  client,
  tutorials,
  feedback
};
