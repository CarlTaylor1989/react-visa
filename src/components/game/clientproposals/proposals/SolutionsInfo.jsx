import React, { Component } from 'react';
import { I18n, Translate } from 'react-redux-i18n';
import PropTypes from 'prop-types';
import ScrollBar from 'react-scrollbars-custom';
import scrollBarSettings from '../../../../config/scrollBar';
import { sendInteractionData } from '../../../../tracking/xapi/XApiAdapter';
import { clientsPropsReferrer } from '../../../../config/referrers';

import './SolutionsInfo.scss';

class SolutionsInfo extends Component {
  constructor(props) {
    super(props);
    this.timeAvailable = null;
    this.onClose = this.onClose.bind(this);
  }

  componentDidUpdate(prevProps) {
    const { showInfo } = this.props;

    if (!prevProps.showInfo && showInfo) {
      this.timeAvailable = new Date();
    }
  }

  componentWillUnmount() {
    const { client, id } = this.props;

    if (client && id && this.timeAvailable) {
      this.sendPopupStatement();
    }
  }

  onClose() {
    const { closeInfo } = this.props;
    this.sendPopupStatement();
    closeInfo();
  }

  sendPopupStatement() {
    const { client, id } = this.props;

    sendInteractionData({
      type: 'experienced',
      activity: {
        id: `${clientsPropsReferrer}/${client}/question-2/plm${id}`,
        type: 'slide',
        title: I18n.t('xapi.clientproposals.learnMoreTitle', {
          product: I18n.t(`clientprops.clients.${client}.products.${id}.title`)
        }),
        description: I18n.t(`clientprops.clients.${client}.products.${id}.text`)
      },
      response: {
        timeAvailable: this.timeAvailable,
        timeResponse: new Date()
      },
      parent: `${clientsPropsReferrer}/${client}/question-2`
    });

    this.timeAvailable = null;
  }

  render() {
    const {
      id,
      client,
      showInfo
    } = this.props;

    return (
      <div className={`solutionsInfo${showInfo ? ' available' : ''}`}>
        <button
          type="button"
          className="exitPopup"
          onClick={this.onClose}
        >
          <Translate value="generic.popupCloseBtn" />
        </button>
        <div className="contentScroll">
          <ScrollBar {...scrollBarSettings}>
            <div className="popupWrapper">
              <Translate
                className="solutionTitle"
                dangerousHTML
                value={`clientprops.clients.${client}.products.${id}.popupTitle`}
              />
              <Translate
                className="solutionSubtitle"
                dangerousHTML
                value={`clientprops.clients.${client}.products.${id}.popupSubtitle`}
              />
              <Translate
                className="solutionText"
                dangerousHTML
                value={`clientprops.clients.${client}.products.${id}.popupText`}
              />
            </div>
          </ScrollBar>
        </div>
      </div>
    );
  }
}

SolutionsInfo.propTypes = {
  id: PropTypes.string.isRequired,
  client: PropTypes.string.isRequired,
  closeInfo: PropTypes.func.isRequired,
  showInfo: PropTypes.bool.isRequired
};

export default SolutionsInfo;
