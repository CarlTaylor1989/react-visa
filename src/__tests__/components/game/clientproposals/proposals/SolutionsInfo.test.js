/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import SolutionsInfo from
  '../../../../../components/game/clientproposals/proposals/SolutionsInfo';
import { sendInteractionData } from '../../../../../tracking/xapi/XApiAdapter';

jest.mock('../../../../../tracking/xapi/XApiAdapter');

let mockProps;
beforeEach(() => {
  mockProps = {
    id: '0',
    type: 'services',
    client: 'client1',
    closeInfo: jest.fn(),
    showInfo: true
  };
});

test('renders a visible SolutionsInfo component', () => {
  const wrapper = shallow(<SolutionsInfo {...mockProps} />, {
    disableLifecycleMethods: true
  });
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders a hidden SolutionsInfo component', () => {
  const wrapper = shallow(<SolutionsInfo {...mockProps} showInfo={false} />, {
    disableLifecycleMethods: true
  });
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

describe('method componentDidUpdate()', () => {
  test('sets the timeAvailable', () => {
    const wrapper = shallow(<SolutionsInfo {...mockProps} showInfo={false} />);
    wrapper.setProps({
      showInfo: true
    });
    expect(wrapper.instance().timeAvailable).toBeDefined();
  });

  test('timeAvailable in not defined', () => {
    const wrapper = shallow(<SolutionsInfo {...mockProps} />);
    wrapper.setProps({
      showInfo: false
    });
    expect(wrapper.instance().timeAvailable).toBeNull();
  });
});

describe('method componentWillUnmount()', () => {
  test('calls the method sendPopupStatement()', () => {
    // setup
    const wrapper = shallow(<SolutionsInfo {...mockProps} />, {
      disableLifecycleMethods: true
    });
    wrapper.instance().timeAvailable = new Date();
    wrapper.instance().sendPopupStatement = jest.fn();

    // function under test
    wrapper.instance().componentWillUnmount();

    // expectation
    expect(wrapper.instance().sendPopupStatement).toHaveBeenCalled();
  });

  test('method sendPopupStatement() is not called', () => {
    // setup
    const wrapper = shallow(<SolutionsInfo {...mockProps} />, {
      disableLifecycleMethods: true
    });
    wrapper.instance().sendPopupStatement = jest.fn();

    // function under test
    wrapper.instance().componentWillUnmount();

    // expectation
    expect(wrapper.instance().sendPopupStatement).not.toHaveBeenCalled();
  });
});

test('method onClose()', () => {
  // setup
  const wrapper = shallow(<SolutionsInfo {...mockProps} />, {
    disableLifecycleMethods: true
  });
  wrapper.instance().sendPopupStatement = jest.fn();

  // function under test
  wrapper.instance().onClose();

  // expectations
  expect(wrapper.instance().sendPopupStatement).toHaveBeenCalled();
  expect(mockProps.closeInfo).toHaveBeenCalled();
});

test('method sendPopupStatement()', () => {
  // setup
  const wrapper = shallow(<SolutionsInfo {...mockProps} />, {
    disableLifecycleMethods: true
  });
  wrapper.instance().timeAvailable = new Date();

  // function under test
  wrapper.instance().sendPopupStatement();

  // expectation
  expect(sendInteractionData).toHaveBeenCalledWith(expect.objectContaining({
    type: 'experienced',
    activity: expect.objectContaining({
      id: expect.stringContaining('question'),
      type: 'slide',
      title: expect.any(String),
      description: expect.any(String)
    }),
    response: expect.objectContaining({
      timeAvailable: expect.any(Date),
      timeResponse: expect.any(Date)
    }),
    parent: expect.any(String)
  }));
});
