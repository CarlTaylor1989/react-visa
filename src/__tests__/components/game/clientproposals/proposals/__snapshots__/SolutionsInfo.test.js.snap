// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a hidden SolutionsInfo component 1`] = `
<div
  className="solutionsInfo"
>
  <button
    className="exitPopup"
    onClick={[Function]}
    type="button"
  >
    <Translate
      tag="span"
      value="generic.popupCloseBtn"
    />
  </button>
  <div
    className="contentScroll"
  >
    <Scrollbar
      contentProps={Object {}}
      fallbackScrollbarWidth={20}
      maximalThumbYSize={83}
      minimalThumbSize={30}
      minimalThumbYSize={83}
      momentum={true}
      noDefaultStyles={true}
      scrollDetectionThreshold={100}
      scrollerProps={Object {}}
      thumbXProps={Object {}}
      thumbYProps={Object {}}
      trackClickBehavior="jump"
      trackXProps={Object {}}
      trackYProps={Object {}}
      wrapperProps={Object {}}
    >
      <div
        className="popupWrapper"
      >
        <Translate
          className="solutionTitle"
          dangerousHTML={true}
          tag="span"
          value="clientprops.clients.client1.products.0.popupTitle"
        />
        <Translate
          className="solutionSubtitle"
          dangerousHTML={true}
          tag="span"
          value="clientprops.clients.client1.products.0.popupSubtitle"
        />
        <Translate
          className="solutionText"
          dangerousHTML={true}
          tag="span"
          value="clientprops.clients.client1.products.0.popupText"
        />
      </div>
    </Scrollbar>
  </div>
</div>
`;

exports[`renders a visible SolutionsInfo component 1`] = `
<div
  className="solutionsInfo available"
>
  <button
    className="exitPopup"
    onClick={[Function]}
    type="button"
  >
    <Translate
      tag="span"
      value="generic.popupCloseBtn"
    />
  </button>
  <div
    className="contentScroll"
  >
    <Scrollbar
      contentProps={Object {}}
      fallbackScrollbarWidth={20}
      maximalThumbYSize={83}
      minimalThumbSize={30}
      minimalThumbYSize={83}
      momentum={true}
      noDefaultStyles={true}
      scrollDetectionThreshold={100}
      scrollerProps={Object {}}
      thumbXProps={Object {}}
      thumbYProps={Object {}}
      trackClickBehavior="jump"
      trackXProps={Object {}}
      trackYProps={Object {}}
      wrapperProps={Object {}}
    >
      <div
        className="popupWrapper"
      >
        <Translate
          className="solutionTitle"
          dangerousHTML={true}
          tag="span"
          value="clientprops.clients.client1.products.0.popupTitle"
        />
        <Translate
          className="solutionSubtitle"
          dangerousHTML={true}
          tag="span"
          value="clientprops.clients.client1.products.0.popupSubtitle"
        />
        <Translate
          className="solutionText"
          dangerousHTML={true}
          tag="span"
          value="clientprops.clients.client1.products.0.popupText"
        />
      </div>
    </Scrollbar>
  </div>
</div>
`;
